torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -0,0 +1,192 @@
+      // For this SyncInserter
+      (!initial_sync_) ? hasOutputSmemExpr(expr, initial_)
+                       : hasInputSmemExpr(expr, final_);
+
+      // For parent SyncInserter
+      hasOutputSmemExpr(expr, all_smem_outputs_);
+      hasInputSmemExpr(expr, all_smem_inputs_);
+    } else {
+      OptOutDispatch::handle(expr);
+    }
+  }
+
+  void handle(kir::IfThenElse* ite) final {
+    for (auto expr : ite->thenBody().exprs()) {
+      handle(expr);
+    }
+    for (auto expr : ite->elseBody().exprs()) {
+      handle(expr);
+    }
+  }
+
+  void handle(kir::ForLoop* fl) final {
+    for (auto expr : fl->body().exprs()) {
+      last_op_sync_ = false;
+      if (expr->getExprType().value() == ExprType::Sync) {
+        initial_sync_ = true;
+        final_.clear();
+      } else if (expr->getExprType().value() == ExprType::ForLoop) {
+        // Recursively handle nested for-loop
+        LocalSyncInserter child(expr);
+        const auto& child_inputs = child.all_smem_inputs_;
+        const auto& child_outputs = child.all_smem_outputs_;
+
+        // Default - Track all smem inputs / outputs
+        all_smem_inputs_.insert(child_inputs.begin(), child_inputs.end());
+        all_smem_outputs_.insert(child_outputs.begin(), child_outputs.end());
+
+        if (!initial_sync_) {
+          // Parent - None
+          if (!child.initial_sync_) {
+            // Child - None
+            // Append All Child Outputs to Parent Initial
+            initial_.insert(child_outputs.begin(), child_outputs.end());
+          } else if (child.war_sync_) {
+            // Child - WAR race
+            // Parent first sync
+            // Inherit Child Initial / Clear Parent Final
+            initial_sync_ = true;
+            last_op_sync_ = true;
+            initial_.insert(child.initial_.begin(), child.initial_.end());
+            final_.clear();
+          } else {
+            // Child - 1+
+            // Parent first sync
+            // Inherit Child Initial + Final
+            initial_sync_ = true;
+            initial_.insert(child.initial_.begin(), child.initial_.end());
+            final_.insert(child.final_.begin(), child.final_.end());
+          }
+        } else {
+          // Parent - 1+
+          if (!child.initial_sync_) {
+            // Child - None
+            // Append All Child to Parent Last
+            final_.insert(child_inputs.begin(), child_inputs.end());
+          } else if (child.war_sync_) {
+            // Child - WAR race
+            // Clear Parent Last / Discard Child Initial
+            last_op_sync_ = true;
+            final_.clear();
+          } else {
+            // Child - 1+
+            // Inherit Child Final / Discard Child Initial
+            final_.insert(child.final_.begin(), child.final_.end());
+          }
+        }
+      } else {
+        handle(expr);
+      }
+    }
+
+    if (!fl->iter_domain()->isThread() && !fl->iter_domain()->isBroadcast()) {
```

Comment:
Why is this check of `isBroadcast()`?

Reply:
https://github.com/csarofeen/pytorch/blob/20_8_18_devel/torch/csrc/jit/codegen/cuda/codegen.cpp#L534-L549

I used this from codegen to determine if the for-loop actually occurs in the kernel.

Reply:
That makes sense. Thanks.

Reply:
@rdspring1, can you please add a comment with that explanation?